src/bson_util/async_decode.rs

```diff
@@ -0,0 +1,214 @@
+    let mut arr = Array::new();
+
+    // disregard the length: using Read::take causes infinite type recursion
+    read_i32(reader).await?;
+
+    loop {
+        let tag = read_u8(reader).await?;
+        if tag == 0 {
+            break;
+        }
+
+        // check that the key is as expected
+        let key = read_cstring(reader).await?;
+        match key.parse::<usize>() {
+            Err(..) => return Err(DecoderError::InvalidArrayKey(arr.len(), key)),
+            Ok(idx) => {
+                if idx != arr.len() {
+                    return Err(DecoderError::InvalidArrayKey(arr.len(), key));
+                }
+            }
+        }
+
+        let val = decode_bson(reader, tag, utf8_lossy).await?;
+        arr.push(val)
+    }
+
+    Ok(arr)
+}
+
+fn decode_bson<R: AsyncRead + Unpin + Send>(
+    reader: &mut R,
+    tag: u8,
+    utf8_lossy: bool,
+) -> BoxFuture<DecoderResult<Bson>> {
+    use bson::spec::{BinarySubtype, ElementType::*};
+
+    Box::pin(async move {
+        match bson::spec::ElementType::from(tag) {
+            Some(FloatingPoint) => Ok(Bson::FloatingPoint(
+                reader.read_f64::<LittleEndian>().await?,
+            )),
+            Some(Utf8String) => read_string(reader, utf8_lossy).await.map(Bson::String),
+            Some(EmbeddedDocument) => decode_document(reader).await.map(Bson::Document),
+            Some(Array) => decode_array(reader, utf8_lossy).await.map(Bson::Array),
+            Some(Binary) => {
+                let len = read_i32(reader).await?;
+                if len < 0 || len > MAX_BSON_SIZE {
+                    return Err(DecoderError::InvalidLength(
+                        len as usize,
+                        format!("Invalid binary length of {}", len),
+                    ));
+                }
+                let subtype = BinarySubtype::from(read_u8(reader).await?);
+                let mut data = Vec::with_capacity(len as usize);
+                reader.take(len as u64).read_to_end(&mut data).await?;
+                Ok(Bson::Binary(subtype, data))
+            }
+            Some(ObjectId) => {
+                let mut objid = [0; 12];
+                for x in &mut objid {
+                    *x = read_u8(reader).await?;
+                }
+                Ok(Bson::ObjectId(oid::ObjectId::with_bytes(objid)))
+            }
+            Some(Boolean) => Ok(Bson::Boolean(read_u8(reader).await? != 0)),
+            Some(NullValue) => Ok(Bson::Null),
+            Some(RegularExpression) => {
+                let pat = read_cstring(reader).await?;
+                let opt = read_cstring(reader).await?;
+                Ok(Bson::RegExp(pat, opt))
+            }
+            Some(JavaScriptCode) => read_string(reader, utf8_lossy)
+                .await
+                .map(Bson::JavaScriptCode),
+            Some(JavaScriptCodeWithScope) => {
+                // disregard the length:
+                //     using Read::take causes infinite type recursion
```

Comment:
what is this comment referring to?

Reply:
It was a pre-existing comment from the BSON implementation. I think they're referring to the fact that using `take` to limit the number of bytes that can be read similar to how it's used elsewhere in the file. I've removed all the comments like this with the most recent commit.

Reply:
I think that commit hasn't been pushed yet.

Reply:
Oops; now it has been pushed